### PR TITLE
Expose structure of annotated methods via ManagedClass

### PR DESCRIPTION
**Pull request by mattstep**

### CODE REVIEW

File: src/main/java/org/weakref/jmx/AnnotationUtils.java [hunk 1]
```diff
@@ -214,6 +214,27 @@ public static String getDescription(Annotation... annotations)
...
+
+    public static String getName(Annotation... annotations)
+    {
+        String name = "";
+        for (Annotation annotation : annotations) {
+            if (annotation instanceof Managed) {
+                try {
+                    Method nameMethod = annotation.annotationType().getMethod("name");
+                    name = nameMethod.invoke(annotation).toString();
+                } catch (ReflectiveOperationException e) {
```
Comment (martint) [inline src/main/java/org/weakref/jmx/AnnotationUtils.java:230]: `catch` should go in the next line

File: src/main/java/org/weakref/jmx/MBeanExporter.java [hunk 2]
```diff
@@ -215,14 +219,23 @@ public Map<String, Exception> unexportAllAndReportMissing()
     public Map<String, Object> getExportedObjects()
     {
         synchronized (exportedObjects) {
-            ImmutableMap.Builder<String,Object> builder = ImmutableMap.builder();
+            ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();
```
Comment (martint) [inline src/main/java/org/weakref/jmx/MBeanExporter.java:222]: Unrelated formatting change. Add it as a separate commit.

File: src/main/java/org/weakref/jmx/ManagedAttribute.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
+/**
+ *  Copyright 2022 Matt Stephenson
```
Comment (martint) [inline src/main/java/org/weakref/jmx/ManagedAttribute.java:2]: Remove the copyright line. It won't be accurate as this class evolves over time and it can always be inferred from the change history.

File: src/main/java/org/weakref/jmx/ManagedClass.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        return target.get() != null;
+    }
+
+    public Class getTargetClass()
+    {
+        if (target.get() == null) {
+            throw new IllegalStateException("Reference to target is no longer present.");
+        }
+        return target.get().getClass();
```
Comment (martint) [inline src/main/java/org/weakref/jmx/ManagedClass.java:93]: The reference can disappear in between the `if` and this line. You need to get a strong reference to the object at the beginning and use that instead:

```java
Object reference = target.get();
if (reference == null) {
    throw ...;
}
return reference.getClass();
```

File: src/main/java/org/weakref/jmx/ManagedClass.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    public Object getTarget()
+    {
+        if (target.get() == null) {
```
Comment (martint) [inline src/main/java/org/weakref/jmx/ManagedClass.java:98]: Same here

File: src/main/java/org/weakref/jmx/ManagedClass.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                    }
+                }
+
+                attributes.put(attributeName, new ManagedAttribute(concreteMethod, attributeName, description, AnnotationUtils.isFlatten(concreteMethod)));
+            }
+        }
+        return new ManagedClass(target, children.build(), attributes.build());
+    }
+
+    public boolean isTargetPresent()
```
Comment (martint) [inline src/main/java/org/weakref/jmx/ManagedClass.java:83]: What's the purpose of this method? Any calls are going to suffer from potential race conditions with the GC, so it doesn't seem as useful.

File: src/main/java/org/weakref/jmx/ManagedClass.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        return ImmutableSet.copyOf(attributes.keySet());
+    }
+
+    public Object invokeAttribute(String attributeName)
+            throws InvocationTargetException, IllegalAccessException
+    {
+        if (target.get() == null) {
+            throw new IllegalStateException("Reference to target is no longer present.");
+        }
+        return getManagedAttribute(attributeName).getMethod().invoke(target.get());
```
Comment (martint) [inline src/main/java/org/weakref/jmx/ManagedClass.java:126]: Same here regarding race condition

**Review by martint — CHANGES_REQUESTED** [review]

File: src/main/java/org/weakref/jmx/ManagedClass.java (hunk 6 above)
Comment (mattstep) [reply src/main/java/org/weakref/jmx/ManagedClass.java:83]: Removed, I believe I already refactored not to use this, I'll make sure of it.

File: src/main/java/org/weakref/jmx/ManagedClass.java (hunk 4 above)
Comment (mattstep) [reply src/main/java/org/weakref/jmx/ManagedClass.java:93]: I refactored getTarget() to do this and reused it in all the other places.

File: src/main/java/org/weakref/jmx/ManagedClass.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    private ManagedClass(Object target, Map<String, ManagedClass> children, Map<String, ManagedAttribute> attributes)
+    {
+        this.target = new WeakReference<>(target);
+        this.children = children;
+        this.attributes = attributes;
+    }
+
+    public static ManagedClass fromExportedObject(Object target)
+    {
+        Preconditions.checkNotNull(target, "target is null");
```
Comment (martint) [inline src/main/java/org/weakref/jmx/ManagedClass.java:47]: This is not really need. The usages below will cause an NPE anyway, and in Java 14+ it will produce a better error message.

File: src/main/java/org/weakref/jmx/ManagedClass.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            if (isGetter(concreteMethod)) { // is it an attribute?
+
+                String attributeName = AnnotationUtils.getName(annotatedMethod);
+                String description = AnnotationUtils.getDescription(annotatedMethod);
+
+                if (attributeName == null || attributeName.equals("")) {
+                    attributeName = getAttributeName(concreteMethod);
+                }
+
+                if (AnnotationUtils.isNested(annotatedMethod)) { // is this nested?
```
Comment (martint) [inline src/main/java/org/weakref/jmx/ManagedClass.java:65]: Comment is redundant
